### PR TITLE
[core/swarm] Refactor and extend configurable connection limits.

### PR DESCRIPTION
**Pull request by romanb**
The initial motivation for the work in this PR is https://github.com/libp2p/rust-libp2p/issues/1839. However, the PR goes a bit further than to just add a single counter to have a more streamlined API. Thus, in order to better track different connection counts and permit configurable limits for these counts and make these available for inspection in an efficient manner (i.e. without linear counting operations), I introduced dedicated connection counters via a `ConnectionCounters` structure maintained by the internal connection pool but exposed on the API via the `NetworkInfo`. All connection states that are counted in this way can also have effective configurable limits.

All limits are now configured through a `ConnectionLimits` structure. The new configurable limits are

  * `ConnectionLimits::max_established_incoming` for limiting the number of concurrent established connections that were incoming.
  * `ConnectionLimits::max_established_outgoing` for limiting the number of concurrent established connections that were outgoing.

A limit that has been removed because tracking the related counts efficiently would be a bit laborious and I don't think it is particularly useful is

  * `NetworkConfig::set_outgoing_per_peer_limit()`

Note that there are still the pre-existing `ConnectionLimits::max_incoming` and `ConnectionLimits::max_outgoing` which correspond to the two new configuration options above, just for pending connections. I also took the liberty to change the connection limits to use `u32` instead of `usize` for their types, which seems more appropriate.

### Fixed panic in the `Peer` API

The `Network::peer()` API allows to iterate mutably over the ongoing connection attempts via `DialingPeer::attempts()`, where each `DialingAttempt` can be `abort()`ed during iteration. However, the iterator did not take into account the changes to the internal iteration offsets when elements are removed. This was discovered while extending the connection limit tests a bit and has been fixed now as a result.

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
Good catch on the iterator state mismatch!

File: core/src/connection/pool.rs [hunk 1]
```diff
@@ -894,62 +887,169 @@ where
...
-    pub max_outgoing_per_peer: Option<usize>,
+/// Network connection information.
+#[derive(Debug, Clone)]
+pub struct ConnectionCounters {
+    /// The effective connection limits.
+    limits: ConnectionLimits,
+    /// The current number of incoming connections.
+    incoming: u32,
+    /// The current number of outgoing connections.
+    outgoing: u32,
```
[the remark this answers is not in the document]
Comment (romanb) [reply core/src/connection/pool.rs:898]: Yes, sounds fine to me.

**Review by tomaka — APPROVED** [review]
Looks good to me with the same remarks as Max on the naming.

**Comment by romanb** [comment]
I incorporated the naming suggestion(s) in https://github.com/libp2p/rust-libp2p/pull/1848/commits/4e447e99a0ec9f740732a6935a639b7c57af23e6 together with a bit refinement for the configuration APIs:

  * `ConnectionLimits` now has private fields for future extensions and offers a builder API that moves `self`.
  * `NetworkConfig` now has a builder API that moves `self`.

The `SwarmBuilder` already used this style, so there is some gain in consistency. Let me know if that last commit looks acceptable.

File: core/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,12 @@
 # 0.25.0 [unreleased]
 
+- The `NetworkConfig` API is now a builder that moves `self`.
```
Comment (mxinden) [inline core/CHANGELOG.md:3]:
```suggestion
- The `NetworkConfig` API is now a builder that moves `self`.
  [PR 1848](https://github.com/libp2p/rust-libp2p/pull/1848/).
```

File: core/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
+
+- New configurable connection limits for established connections and
+  dedicated connection counters. Removed the connection limit dedicated
+  to outgoing pending connection _per peer_. Connection limits are now
+  represented by `u32` intead of `usize` types.
```
Comment (mxinden) [inline core/CHANGELOG.md:8]:
```suggestion
  represented by `u32` intead of `usize` types.
  [PR 1848](https://github.com/libp2p/rust-libp2p/pull/1848/).
```

**Review by mxinden — APPROVED** [review]
Both the connection limiting as well as the general API improvements look good to me.

**Pull request merged**